### ISSUE TITLE
layer.toDataURL returning empty string on Safari 15

### ISSUE THREAD

**Issue by richardjonlewis** [konvajs/konva#1290]
Hi All, I believe I've come across a bug with Konva on Safari 15.  Essentially the layer.toDataURL returns an empty string "" for some layer scale and toDataURL pixelRatio combinations, even when well within canvas size limits (i.e. I'm not talking about when .toDataURL() returns "data:;".  This returns a totally empty string "").

I've created jsFiddle below. This works on all browsers I've tested it on (various Chrome, Firefox versions and Safari 13 and 14) but fails on Safari 15. The fiddle draws a layer with some circles in. Then cycles through some arbitrary zoom factors (layer scales) and for each it then tries various pixel ratios making sure the total canvas size never goes above 153,360,000 (8000x19200) pixels or below width or height of <10 pixels.  And for each of these combinations, it fires layer.toDataURL and checks if the resulting string is < 10 characters.  If it finds any it lists them above the canvas.

[https://jsfiddle.net/richardjonlewis/6ho529xk/76](https://jsfiddle.net/richardjonlewis/6ho529xk/76)

For working browsers it will find no dataURLs of < 10 chars. like screenshot below (Safari 14). 
![iPadSafari14](https://user-images.githubusercontent.com/10913175/155411031-fc779b82-2470-44c5-ab4f-7acc05155547.jpeg)

For Safari 15 it will list the scale and pixelRatio combinations that resulted in a string of less than 10 chars (or empty string). Screenshot of Safari 15 (hidden in screenshot are other listed scale and pixelRatio combinations that result in empty string from toDataURL). It does create a full dataURL for some combinations but some combinations result in an empty string.
![iPhoneSafari15](https://user-images.githubusercontent.com/10913175/155411056-a87526ea-2dd9-4fa3-9100-ad732712fd54.jpeg)

Edit: I should add that my error logs suggest this is happening across all Safari 15 versions on mac, Ipad and iPhone.

Thanks in advance for any advice thoughts and for all hard work and the awesome konva library.

**Comment by richardjonlewis** [comment]
I should add that my error logs suggest this is happening across all Safari 15 versions on mac, Ipad and iPhone.

**Comment by lavrton** [comment]
I don't think there is anything we can do from Konva side. This is how Safari works.

Also I don't understand what the demo is doing. Remember that `pixelRatio` is making direct effect on size of canvas used to make an export. 

So if you have `width = 100` and `pixelRatio = 0.01`, Konva will create canvas with `width = 100 * 0.01 = 1` to make an export. Looks like with some of your values it may go < 0.

**Comment by richardjonlewis** [comment]
Thanks @lavrton.  Appreciate the quick response.  So when you say that's how safari works do you mean it returns blank string on <0 width (and height) or does it have a higher lower limit? 

The example I coded makes sure the resultant image doesn't go below 10 width (so 10x24). But the example in the screenshot below returns "" where the resulting image dimensions would be 89 x 215  = 500 * 0.406 * (1/2.26) x 1200 * 0.406 * (1/2.26).  And my actual code makes sure the resultant image will never be <120 in any one direction.  But it still returns an empty string on some occasions (all of which are well above that 120 min limit and below the max canvas size limits).

As for why I created the demo like it did.  In my actual code, I use konva to represent a physical layout with two layers (main and activeLayer). I use scaling of these layers to zoom in and drag this layout representation and I use toDataURL on any change to the layout to speed up zooming or moving it  The dataURL img is then moved and zoomed on an "activeLayer" that is separate from the main layer to improve performance.  Because of this and so that I don't have to change the main layer scale before I snapshot it with toDataURL I use the following params:
```
const dataURL = layer.toDataURL({
    x: layer.x(),
    y: layer.y(),
    width: physicalLayoutWidth * layerScale, //snapshot full layout.
    height: physicalLayoutHeight * layerScale,
    pixelRatio: desiredRatio * (1/layerScale), //to bring image size back down to desired size
});
```
So the demo essentially zooms in, then takes a load of toDataURLs at different pixelRatios then zooms in some more and does the same. and so on.   Making sure the resultant image never gets below that 10 width. 

The `while (500*pixelRatio > 10){}` makes sure that the resulting image size won't get below 10 px width (smallest dimension).

**Comment by richardjonlewis** [comment]
After some research and a lot of testing, I believe it is a bug in Safari with how it manages garbage collection associated with canvas stuff. Many people are having a similar issue with video capture from the canvas.

The approach I was taking was if the toDataURL returned a blank string I would try to reduce/increase the pixelRatio in the hope that it was a max/min canvas size issue relating to that toDataURL action I was taking. 

But it looks like, because I was calling toDataURL in the same "thread" as the last time I called it (where it failed for actual max/min size issue), Safari wasn't resetting itself and all subsequent calls in that process / callstack toDataURL would fail no matter what pixelRatio / canvas size was used.  

I found that if I separate the next call from the current process with setTimeout(, >200) it would successfully capture the dataURL.  

I'll close the issue but it is one to keep in mind for anyone firing toDataURL more than once consecutively on Safari 15.
